### ISSUE TITLE
Reproducing Vanilla NAT Baseline

### ISSUE THREAD

**Issue by stephanpeitz** [chenyangh/DSLP#2]
Hi all,

thanks for sharing your code!

I would like to be able to reproduce the Vanilla NAT Baseline (21.18 BLEU on WMT'14 EN-DE).
What is the corresponding command?

Btw, the command for Vanilla NAT with DSLP

`python3 train.py data-bin/wmt14.en-de_kd --source-lang en --target-lang de  --save-dir checkpoints  --eval-tokenized-bleu \
   --keep-interval-updates 5 --save-interval-updates 500 --validate-interval-updates 500 --maximize-best-checkpoint-metric \
   --eval-bleu-remove-bpe --eval-bleu-print-samples --best-checkpoint-metric bleu --log-format simple --log-interval 100 \
   --eval-bleu --eval-bleu-detok space --keep-last-epochs 5 --keep-best-checkpoints 5  --fixed-validation-seed 7 --ddp-backend=no_c10d \
   --share-all-embeddings --decoder-learned-pos --encoder-learned-pos  --optimizer adam --adam-betas "(0.9,0.98)" --lr 0.0005 \ 
   --lr-scheduler inverse_sqrt --stop-min-lr 1e-09 --warmup-updates 10000 --warmup-init-lr 1e-07 --apply-bert-init --weight-decay 0.01 \
   --fp16 --clip-norm 2.0 --max-update 300000  --task translation_lev --criterion nat_loss --arch nat_sd --noise full_mask \ 
   --src-upsample-scale 2 --use-ctc-decoder --ctc-beam-size 1  --concat-yhat --concat-dropout 0.0  --label-smoothing 0.1 \ 
   --activation-fn gelu --dropout 0.1  --max-tokens 8192`

does not work.
I get: `train.py: error: unrecognized arguments: --src-upsample-scale 2 --use-ctc-decoder --ctc-beam-size 1`

Cheers,
Stephan

**Comment by chenyangh** [comment]
Hello Stephan. 

To reproduce the results of Vanilla NAT, you need the following specific configuration (comparing with those given by the [official Fairseq repo](https://github.com/pytorch/fairseq/tree/main/examples/nonautoregressive_translation)):
1. 128 K token per batch, I.e., make sure  `num_of_cards_per_machine * num_machines * update_freq * max_tokens = 128K`
2. 0.1 dropout rate
3. `gleu` activation
`gleu` is not a significant factor, but our experiments showed it was somehow necessary to get a good result (>25 on WMT'14 EN-DE) on the GLAT model. Therefore, we kept `gelu` for all experiments. 


Regarding the error, just remove `--src-upsample-scale 2 --use-ctc-decoder --ctc-beam-size 1`, those are for CTC-based models. I will fix it in the README, thanks.

Bests,
Chenyang

**Comment by stephanpeitz** [comment]
Thanks for quick reply! 
I will give it a try and come back to you.

Cheers,
Stephan

**Comment by chenyangh** [comment]
That being said, you need a script like this,
```bash
python3 train.py data-bin/wmt14.en-de_kd --source-lang en --target-lang de  --save-dir checkpoints  --eval-tokenized-bleu \
   --keep-interval-updates 5 --save-interval-updates 500 --validate-interval-updates 500 --maximize-best-checkpoint-metric \
   --eval-bleu-remove-bpe --eval-bleu-print-samples --best-checkpoint-metric bleu --log-format simple --log-interval 100 \
   --eval-bleu --eval-bleu-detok space --keep-last-epochs 5 --keep-best-checkpoints 5  --fixed-validation-seed 7 --ddp-backend=no_c10d \
   --share-all-embeddings --decoder-learned-pos --encoder-learned-pos  --optimizer adam --adam-betas "(0.9,0.98)" --lr 0.0005 \ 
   --lr-scheduler inverse_sqrt --stop-min-lr 1e-09 --warmup-updates 10000 --warmup-init-lr 1e-07 --apply-bert-init --weight-decay 0.01 \
   --fp16 --clip-norm 2.0 --max-update 300000  --task translation_lev --criterion nat_loss --arch nat --noise full_mask \ 
   --label-smoothing 0.1 \ 
   --activation-fn gelu --dropout 0.1  --max-tokens 8192 
```

But do adjust the batch size according to your computational resource.

**Comment by stephanpeitz** [comment]
Thanks for the full command!

Two questions though:

- You don't do any length prediction as suggested in the official Fairseq repo? E.g. ` --pred-length-offset --length-loss-factor 0.1`
- If you don't, what's the difference between your and Fairseq's vanilla NAT implementation?
- Do you mind to share also your Vanilla NAT hypothesis? 

Thanks a lot!

Cheers,
Stephan

**Comment by chenyangh** [comment]
Thanks for bringing this up, Stephan.
1. I just checked my log and I actually used `--pred-length-offset --length-loss-factor 0.1` in my experiments. The script I gave was adapted from the CTC, and there was some mismatch. 
2. The main difference is the dropout rate. 
- For WMT'14  EN-De, dropout 0.1 gives me 21.11; but gives me 19.68 for dropout 0.3.
- For WMT'16 EN-RO, dropout 0.3 works much better.
- In our paper, we chose the best setting for every task. I.e., 0.1 for EN-DE, 0.3 for EN-RO.
3. I didn't keep the output, but I will try to get a copy for you when I have more time.
Note that the dataset is already compound-splited, and we only perform the standard tokenized BLEU with FairSeq. We did not use the `scripts/compound_split_bleu.sh` again.

**Comment by chenyangh** [comment]
Also, you need to have a 128K batch size and a 300K maximum update.

**Comment by stephanpeitz** [comment]
Hi Chenyang,

thanks for sharing all these details! 
That's very helpful!

I will let you know once I hit 21.18 BLEU. :)

Cheers,
Stephan

**Comment by stephanpeitz** [comment]
I think we are now able to reproduce the baseline, thanks again!
